File: Valid Parentheses.cpp
#include<stack>
class Solution {
public:
    bool isValid(string s) {
        stack<char> shown;
        int size=s.size();
        if(size==0) return true;
        if(s[0]==')' || s[0]=='}' || s[0]==']') return false;
        shown.push(s[0]);
        
        for(int i=1;i<size;i++){
            if(s[i]==')' || s[i]=='}' || s[i]==']'){
                //need to consider when the stack is empty
                if(shown.empty()) return false;
				if((s[i]==')'&&shown.top()=='(')||(s[i]=='}'&&shown.top()=='{')||(s[i]==']'&&shown.top()=='[')) shown.pop();
                else return false;
            }
            else{
                shown.push(s[i]);
            }
        }
        
        if(shown.empty()) return true;
        else return false;
    }
};

File: LeetCode/20. Valid Parentheses/valid_parentheses.py
# just use stack and pay attention to details
class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s) == 0:
            return True
        stack = []
        left = ['(', '[', '{']
        for i in s:
            # handle left
            if i in left:
                stack.append(i)
                continue
                # handle right
            if len(stack) == 0:
                return False
            prev = stack.pop()
            if (prev == '(' and i == ')') or (prev == '[' and i == ']') or (prev == '{' and i == '}'):
                continue
            return False
        if len(stack) > 0:
            return False
        return True